File: src/econode25/projection.py
from __future__ import annotations

from math import cos, pi
from typing import Iterable


METERS_PER_DEG_LAT = 110_540.0


def local_xy_to_lonlat(center_lat: float, center_lon: float, x_m: float, y_m: float) -> tuple[float, float]:
    lon_scale = 111_320.0 * max(cos(center_lat * pi / 180.0), 1e-6)
    lon = center_lon + x_m / lon_scale
    lat = center_lat + y_m / METERS_PER_DEG_LAT
    return lon, lat


def lonlat_to_local_xy(center_lat: float, center_lon: float, lon: float, lat: float) -> tuple[float, float]:
    lon_scale = 111_320.0 * max(cos(center_lat * pi / 180.0), 1e-6)
    x_m = (lon - center_lon) * lon_scale
    y_m = (lat - center_lat) * METERS_PER_DEG_LAT
    return x_m, y_m


def transform_coords(
    center_lat: float,
    center_lon: float,
    coords: Iterable[tuple[float, float]],
) -> list[tuple[float, float]]:
    return [local_xy_to_lonlat(center_lat, center_lon, x, y) for x, y in coords]

File: econode25_map_viewer/src/data/projection.ts
export type Position = [number, number];

const METERS_PER_DEG_LAT = 110540;

export function localXYToLonLat(centerLat: number, centerLon: number, xMeters: number, yMeters: number): Position {
  const lonScale = 111320 * Math.max(Math.cos((centerLat * Math.PI) / 180), 1e-6);
  return [centerLon + xMeters / lonScale, centerLat + yMeters / METERS_PER_DEG_LAT];
}

export function lonLatToLocalXY(centerLat: number, centerLon: number, lon: number, lat: number): Position {
  const lonScale = 111320 * Math.max(Math.cos((centerLat * Math.PI) / 180), 1e-6);
  return [(lon - centerLon) * lonScale, (lat - centerLat) * METERS_PER_DEG_LAT];
}

export function projectGeometry(
  geometry: GeoJSON.Geometry,
  centerLat: number,
  centerLon: number
): GeoJSON.Geometry {
  if (geometry.type === "Point") {
    const [x, y] = geometry.coordinates as [number, number];
    return { type: "Point", coordinates: localXYToLonLat(centerLat, centerLon, x, y) };
  }
  if (geometry.type === "LineString") {
    return {
      type: "LineString",
      coordinates: geometry.coordinates.map(([x, y]) => localXYToLonLat(centerLat, centerLon, x, y))
    };
  }
  if (geometry.type === "Polygon") {
    return {
      type: "Polygon",
      coordinates: geometry.coordinates.map((ring) =>
        ring.map(([x, y]) => localXYToLonLat(centerLat, centerLon, x, y))
      )
    };
  }
  return geometry;
}
